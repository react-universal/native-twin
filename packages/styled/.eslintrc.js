// @generated by expo-module-scripts
module.exports = {
  extends: ['expo-module-scripts/eslintrc.base.js', '../../.eslintrc.js'],
};
