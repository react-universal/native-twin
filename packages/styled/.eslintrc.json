{
  "extends": "../../.eslintrc.json",
  "ignorePatterns": ["!**/*", "build"],
  "overrides": [
    {
      "files": [
        "*.ts",
        "*.tsx"
      ],
      "extends": [
        "plugin:@nx/typescript",
        "plugin:@nx/react"
      ],
      "rules": {
        "@typescript-eslint/no-empty-object-type": "off",
        "@typescript-eslint/no-unsafe-function-type": "off",
        "@typescript-eslint/no-empty-interface": "off",
        "@typescript-eslint/ban-types": "off",
        "@typescript-eslint/no-explicit-any": "off",
        "@typescript-eslint/no-unnecessary-type-constraint": "off",
        "@typescript-eslint/no-non-null-assertion": "off",
        "@typescript-eslint/ban-ts-comment": "off",
        "eqeqeq": "off",
        "@typescript-eslint/no-var-requires": "off"
      }
    }
  ]
}