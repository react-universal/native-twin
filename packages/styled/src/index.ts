export * from './styled';

export { useBuildStyledComponent } from './hooks/useStyledComponent';
export { useStyledContext } from './hooks/useStyledContext';

export * from './styled/variants';

export type {
  PropsFrom,
  StyledComponentProps,
  RegisteredComponent,
  ComponentStylesheet,
} from './types/styled.types';
export type { StyledContext, Units } from './types/css.types';

// CUSTOM
export { createStyledContext } from './utils/createStyledContext';
