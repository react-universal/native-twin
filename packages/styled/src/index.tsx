import { ComponentType, forwardRef, useId, ComponentProps } from 'react';
import { useComponentRegistration, useInteraction } from './hooks';
import type { IExtraProperties } from './styled.types';

function styled<T extends ComponentType>(Component: T) {
  const Styled = forwardRef<T>(({ className = '', children, ...restProps }: any, ref) => {
    const componentID = useId();
    const { style } = useComponentRegistration(componentID, className, restProps?.style);
    const interaction = useInteraction(componentID, restProps);
    console.log('STYLE: ', style);
    return (
      <Component style={style} key={componentID} {...restProps} {...interaction} ref={ref}>
        {children}
      </Component>
    );
  });
  if (typeof Component !== 'string') {
    Styled.displayName = `StyledTW.${Component.displayName || Component.name || 'NoName'}`;
  }
  return Styled as ComponentType<ComponentProps<typeof Component> & IExtraProperties>;
}

export { styled };
export { mergeTWClasses } from './utils/mergeClasses';
