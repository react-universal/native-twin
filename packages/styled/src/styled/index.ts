export { styled } from './StyledComponent';
export type { PropsFrom } from './StyledComponent';
