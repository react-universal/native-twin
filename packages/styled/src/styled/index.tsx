import React from 'react';
import * as RN from 'react-native';
import { LI, UL } from '../custom-components/List';
import {
  styledFlatList,
  styledSectionList,
  styledVirtualizedList,
} from '../custom-components/NativeLists/List.components';
import { Nav } from '../custom-components/Nav';
import { TBody, TD, TFoot, TH, THead, TR, Table } from '../custom-components/Table';
import { Span, H1, H2, H3, H4, H5, H6, Strong, Code, P } from '../custom-components/Text';
import createStyledComponent from './StyledComponent';

const styled = <StyleType, InitialProps extends { style?: RN.StyleProp<StyleType> }>(
  Component: React.ComponentType<InitialProps>,
) => createStyledComponent<StyleType, InitialProps>(Component);

styled.ActivityIndicator = styled(RN.ActivityIndicator);
styled.DrawerLayoutAndroid = styled(RN.DrawerLayoutAndroid);
styled.Image = styled(RN.Image);
styled.ImageBackground = styled(RN.ImageBackground);
styled.KeyboardAvoidingView = styled(RN.KeyboardAvoidingView);
styled.Modal = styled(RN.Modal);
styled.ScrollView = styled(RN.ScrollView);
styled.Switch = styled(RN.Switch);
styled.RefreshControl = styled(RN.RefreshControl);
styled.SafeAreaView = styled(RN.SafeAreaView);
styled.Text = styled(RN.Text);
styled.TextInput = styled(RN.TextInput);
styled.TouchableHighlight = styled(RN.TouchableHighlight);
styled.TouchableNativeFeedback = styled(RN.TouchableNativeFeedback);
styled.TouchableOpacity = styled(RN.TouchableOpacity);
styled.Pressable = styled(RN.Pressable);
styled.TouchableWithoutFeedback = styled(RN.TouchableWithoutFeedback);
styled.View = styled(RN.View);
styled.FlatList = styledFlatList;
styled.SectionList = styledSectionList;
styled.VirtualizedList = styledVirtualizedList;

// WEB
styled.LI = styled(LI);
styled.UL = styled(UL);
styled.Nav = styled(Nav);
styled.Table = styled(Table);
styled.THead = styled(THead);
styled.TFoot = styled(TFoot);
styled.TH = styled(TH);
styled.TD = styled(TD);
styled.TBody = styled(TBody);
styled.TR = styled(TR);

styled.Span = styled(Span);
styled.H1 = styled(H1);
styled.H2 = styled(H2);
styled.H3 = styled(H3);
styled.H4 = styled(H4);
styled.H5 = styled(H5);
styled.H6 = styled(H6);
styled.Strong = styled(Strong);
styled.Code = styled(Code);
styled.P = styled(P);

export default styled;
