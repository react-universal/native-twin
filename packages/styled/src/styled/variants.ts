import type { OmitUndefined, StyledComponentProps } from '../types/styled.types';

type StringToBoolean<T> = T extends 'true' | 'false' ? boolean : T;

type ConfigSchema = Record<string, Record<string, string>>;

export type VariantProps<Component extends (...args: any) => any> = Omit<
  OmitUndefined<Parameters<Component>[0]>,
  'tw' | 'className'
>;

export type VariantsConfig<T = unknown> = T extends ConfigSchema
  ? {
      base?: string;
      variants?: T;
      defaultVariants?: ConfigVariants<T>;
    }
  : never;

export type ConfigVariants<T> = T extends ConfigSchema
  ? {
      [Variant in keyof T]?: StringToBoolean<keyof T[Variant]> | null;
    }
  : unknown;

type VariantsFnProps<T> = T extends ConfigSchema
  ? ConfigVariants<T> & StyledComponentProps
  : StyledComponentProps;

export const createVariants = <T>(config: VariantsConfig<T>) => {
  return (props?: VariantsFnProps<T>) => {
    if (!config || !config?.variants || config.variants == null) {
      let composed = config.base ?? '';
      if (props?.className) {
        composed += `${composed == '' ? '' : ' '}${props.className}`;
      }
      if (props?.tw) {
        composed += `${composed == '' ? '' : ' '}${props.tw}`;
      }
      return composed.trim();
    }
    const { variants, defaultVariants } = config;
    const getVariantClassNames = Object.keys(variants)
      .map((variant: keyof typeof variants) => {
        const variantProp = props?.[variant as keyof typeof props];
        // @ts-expect-error
        const defaultVariantProp = defaultVariants?.[variant];

        if (variantProp === null) return null;

        const variantKey = (falsyToString(variantProp) ||
          falsyToString(defaultVariantProp)) as keyof (typeof variants)[typeof variant];

        return variants[variant]![variantKey];
      })
      .filter(Boolean);
    let composed = config.base ?? '';
    if (getVariantClassNames) {
      composed += `${composed == '' ? '' : ' '}${getVariantClassNames}`;
    }
    if (props?.className) {
      composed += `${composed == '' ? '' : ' '}${props.className}`;
    }
    if (props?.tw) {
      composed += `${composed == '' ? '' : ' '}${props.tw}`;
    }
    return composed.trim();
  };
};

const falsyToString = <T extends unknown>(value: T) =>
  typeof value === 'boolean' ? `${value}` : value === 0 ? '0' : value;
