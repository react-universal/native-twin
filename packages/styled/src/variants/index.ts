import type { ClassProp } from '../types/styledNew';

/**
 * This was forked from https://github.com/joe-bell/cva
 * License: https://github.com/joe-bell/cva/blob/4039155edfd5007cde0e1c9e0060ed838419f242/LICENSE
 */
type ClassValue = string | null | undefined | ClassValue[];
type StringToBoolean<T> = T extends 'true' | 'false' ? boolean : T;

export type ConfigSchema = Record<string, Record<string, ClassValue>>;

export type ConfigVariants<T> = T extends ConfigSchema
  ? {
      [Variant in keyof T]?: StringToBoolean<keyof T[Variant]> | null;
    }
  : unknown;

export type CompoundVariant<T> =
  | (ConfigVariants<T> & ClassProp)
  | ((props?: Props<T>) => string);

export type VariantsConfig<T = unknown> = T extends ConfigSchema
  ? ClassProp & {
      variants?: T;
      defaultProps?: ConfigVariants<T>;
      compoundVariants?: Array<CompoundVariant<T>>;
    }
  : ClassProp;

export type VariantProps<T> = T extends (props?: infer P) => string
  ? Omit<P, keyof ClassProp>
  : never;

type Props<T> = T extends ConfigSchema ? ConfigVariants<T> & ClassProp : ClassProp;

type VariantsFunction<T> = (props?: Props<T>) => string;

export type Variants = {
  <T>(config: VariantsConfig<T>): VariantsFunction<T>;
  <T>(base: string | string[], config?: VariantsConfig<T>): VariantsFunction<T>;
};

export const joinClasses = (classValue: ClassValue): string => {
  // If this is any higher Typescript complains that Array.flat is too deep.
  // If you need more than 21 levels, please raise an issue
  const MAX_SAFE_TYPESCRIPT_FLAT_VALUE = 21;

  return Array.isArray(classValue)
    ? classValue.flat(MAX_SAFE_TYPESCRIPT_FLAT_VALUE).filter(Boolean).join(' ')
    : classValue ?? '';
};

export const variants: Variants =
  <T>(baseOrConfig: string | string[] | VariantsConfig<T>, config?: VariantsConfig<T>) =>
  (props?: Props<T>): string => {
    let base: ClassValue;

    if (typeof baseOrConfig === 'object' && !Array.isArray(baseOrConfig)) {
      config = baseOrConfig;
      base = baseOrConfig.tw ?? baseOrConfig.className;
    } else {
      base = baseOrConfig;
    }

    const variantClassValue: ClassValue = [];
    const propClassValue: ClassValue = props?.tw ?? props?.className;

    if (!config) {
      return joinClasses([base, variantClassValue, propClassValue]);
    }

    if (!('variants' in config) && !('compoundVariants' in config)) {
      return joinClasses([base, variantClassValue, propClassValue]);
    }

    const { variants, defaultProps, compoundVariants } = config;

    const mergedProps: Record<string, unknown> = {
      ...defaultProps,
      ...props,
    };

    if (variants && typeof variants === 'object') {
      for (const variant of Object.keys(variants)) {
        const value = mergedProps?.[variant as keyof typeof props];

        if (value === null || !(variant in variants)) continue;

        const key = value?.toString();

        if (value && key && variants[variant]?.[key]) {
          variantClassValue.push(variants[variant]?.[key]);
        } else if (value && variants[variant]?.['true']) {
          variantClassValue.push(variants[variant]?.['true']);
        } else if (!value && variants[variant]?.['false']) {
          variantClassValue.push(variants[variant]?.['false']);
        }
      }
    }

    if (!compoundVariants) {
      return joinClasses([base, variantClassValue, propClassValue]);
    }

    for (const compoundVariant of compoundVariants) {
      if (typeof compoundVariant === 'function') {
        const match = compoundVariant(props);
        if (match) variantClassValue.push(match);
      } else {
        const { className, tw, ...criteria } = compoundVariant;
        const match = Object.entries(criteria).every(([key, value]) => {
          return typeof value === 'boolean'
            ? Boolean(mergedProps[key]) === value
            : mergedProps[key] === value;
        });

        if (match) variantClassValue.push(tw ?? className ?? '');
      }
    }

    return joinClasses([base, variantClassValue, propClassValue]);
  };
