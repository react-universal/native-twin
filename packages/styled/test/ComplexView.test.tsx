import renderer from 'react-test-renderer';
import { defineConfig, setup } from '@native-twin/native-twin';
import { presetTailwind } from '@native-twin/preset-tailwind';
import { ScrollView, styled, View } from '../src';
import { createVariants } from '../src/styled/variants';

beforeAll(() => {
  setup(defineConfig({ presets: [presetTailwind()] }));
});

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();
  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);
  return result as renderer.ReactTestRendererJSON;
}

describe('@native-twin/styled', () => {
  it('ScrollView render', () => {
    const component = renderer.create(
      <ScrollView className='flex-1'>
        <View />
      </ScrollView>,
    );
    let tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});

const viewVariants = createVariants({
  base: `
    flex-1
    hover:(bg-blue-600)
    p-14 bg-rose-200 border-white border-2
    items-center justify-center md:border-3
  `,
  variants: {
    intent: {
      primary: '',
      secondary: 'm-2',
      third: '',
    },
    active: {
      true: '',
      false: '',
    },
  },
  defaultVariants: {
    active: true,
  },
});
const H1 = styled.Text;

describe('@native-twin/styled', () => {
  it('Complex View', () => {
    const className = viewVariants({ intent: 'secondary' });
    const component = renderer.create(
      <View className={className}>
        <H1 className='text(center 2xl indigo-600) hover:text-gray-700'>H1 - 1</H1>
      </View>,
    );
    let tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});

// NOT WORKING
// describe('@native-twin/styled', () => {
//   it('FlatList', () => {
//     const className = viewVariants({ intent: 'secondary' });
//     const component = renderer.create(
//       <FlatList
//         data={[{ key: 'i1' }, { key: 'i2' }]}
//         renderItem={(props) => (
//           <H1 className='text(center 2xl indigo-600) hover:text-gray-700'>{props.item.key}</H1>
//         )}
//         className={className}
//         keyExtractor={(i) => i.key}
//       />,
//     );
//     let tree = toJson(component);
//     expect(tree).toMatchSnapshot();
//   });
// });
