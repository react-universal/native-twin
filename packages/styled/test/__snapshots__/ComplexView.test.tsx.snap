// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@native-twin/styled Complex View 1`] = `
<View
  groupID=":r2:"
  onBlur={[Function]}
  onFocus={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "rgba(254,205,211,1)",
      "borderColor": "rgba(255,255,255,1)",
      "borderWidth": 2,
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
      "justifyContent": "center",
      "margin": 8,
      "padding": 56,
    }
  }
>
  <Text
    groupID=":r2:"
    onBlur={[Function]}
    onFocus={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    parentID=":r2:"
    style={
      {
        "color": "rgba(79,70,229,1)",
        "fontSize": 24,
        "textAlign": "center",
      }
    }
  >
    H1 - 1
  </Text>
</View>
`;

exports[`@native-twin/styled ScrollView render 1`] = `
<RCTScrollView
  groupID=":r0:"
  style={
    {
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <View>
    <View
      groupID=":r0:"
      parentID=":r0:"
      style={{}}
    />
  </View>
</RCTScrollView>
`;
